### PR TITLE
feat: add module "operation" to support defining the change of a draft

### PR DESCRIPTION
**Pull request by graczhual**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-28v2f90 实现 Draft 级别的 ActionList](https://app.clickup.com/t/28v2f90)

File: graviti/manager/draft.py [hunk 1]
```diff
@@ -57,6 +59,23 @@ def __init__(
         self.created_at = created_at
         self.updated_at = updated_at
         self.description = description
+        self.operations: Set[Union[SheetPost, SheetGet, SheetDelete]] = set()
+
+    def __getitem__(self, key: str) -> DataFrame:
+        value = self._get_data().__getitem__(key)
```
Comment (linjiX) [inline graviti/manager/draft.py:65]: `super().__getitem__(xxx)?`

File: graviti/openapi/sheet.py [hunk 2]
```diff
@@ -68,14 +69,8 @@ def create_sheet(
...
         ...     name = "val",
-        ...     schema = "{\"imports\": [{\"repo\":
-        ...         \"https://github.com/Project-OpenBytes/standard@main\", \"types\":
-        ...         [{\"name\": \"label.Category\"}, {\"name\": \"tensor.Image\"}]}],
-        ...         \"type\": \"record\", \"fields\": [{\"name\": \"image\", \"type\":
-        ...         \"tensor.Image\", \"dtype\": \"uint8\", \"shape\": [28, 28]}, {\"name\":
-        ...         \"category\", \"type\": \"label.Category\", \"categories\": [0, 1, 2, 3,
-        ...         4, 5, 6, 7, 8, 9]}]}",
-        ...     _arrow_schema = ""
+        ...     schema = "",
```
Comment (linjiX) [inline graviti/openapi/sheet.py:72]: why set the example schema to an empty string?

File: graviti/operation/draft.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+
+"""Definitions of different operations about the sheet on a draft."""
+
+from graviti.openapi import create_sheet, delete_sheet
+
+
+class DraftOperation:
```
Comment (linjiX) [inline graviti/operation/draft.py:11]: will `SheetOperation` better？since all the subclass are sheet related.

File: graviti/operation/draft.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    """This class defines the operation that create a sheet.
+
+    Arguments:
+        sheet: The sheet name.
+
+    """
+
+    def __init__(self, sheet: str) -> None:
+        super().__init__(sheet)
+        self.func = create_sheet
```
Comment (linjiX) [inline graviti/operation/draft.py:33]: Is `self.func` a constant? should we set func as a class variable?

File: graviti/manager/draft.py [hunk 5]
```diff
@@ -57,6 +59,23 @@ def __init__(
...
+        if SheetPost(key) not in self.operations:
+            self.operations.add(SheetGet(key))
```
Comment (linjiX) [inline graviti/manager/draft.py:67]: why we need to add operations for `__getitem__`, `__getitem__` should be a read-only move?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/manager/draft.py (hunk 5 above)
Comment (graczhual) [reply graviti/manager/draft.py:67]: We need to know which sheets may have been modified. Then we don't need to go through all the sheets.

File: graviti/manager/draft.py [hunk 6]
```diff
@@ -57,6 +59,23 @@ def __init__(
         self.created_at = created_at
         self.updated_at = updated_at
         self.description = description
+        self.operations: Set[Union[SheetPost, SheetGet, SheetDelete]] = set()
```
Comment (linjiX) [inline graviti/manager/draft.py:62]: python set is unordered, does the operation order matter?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 7]
```diff
@@ -89,6 +90,7 @@ def __init__(
 
         self._columns = {}
         self._column_names = []
+        self.actions: List[Union[DataPatch, DataPost]] = []
```
Comment (linjiX) [inline graviti/dataframe/frame.py:93]: `actions`？or `operations`

File: graviti/manager/draft.py [hunk 8]
```diff
@@ -57,6 +59,15 @@ def __init__(
         self.created_at = created_at
         self.updated_at = updated_at
         self.description = description
+        self.operations: List[Union[SheetPost, SheetDelete]] = []
+
+    def __setitem__(self, key: str, value: DataFrame) -> None:
```
Comment (linjiX) [inline graviti/manager/draft.py:64]: setitem can not only represent `create` but also `replace`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/operation/frame.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    """This class defines the operation that add data to a DataFrame.
+
+    Arguments:
+        data: The data to be added.
+        strategy_argument: The dict about the argument of the Primary
+            key generation strategy of the sheet.
+
+    """
+
+    FUNC = staticmethod(add_data)
```
Comment (linjiX) [inline graviti/operation/frame.py:23]: FUNC is not a good name, will the name `do` be better?

File: graviti/operation/frame.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+#
+
+"""Definitions of different operations on a DataFrame."""
+
+from typing import Any, Dict, List, Optional
+
+from graviti.openapi import add_data, update_data
+
+
+class DataPost:
```
Comment (linjiX) [inline graviti/operation/frame.py:13]: upload data is quite complicated, binary data should be uploaded to oss/s3, also we need multithread to enhance the real time performance. you should consider that.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/operation/frame.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    """This class defines the operation that add data to a DataFrame.
+
+    Arguments:
+        data: The data to be added.
+        strategy_argument: The dict about the argument of the Primary
+            key generation strategy of the sheet.
+
+    """
+
+    DO = staticmethod(add_data)
```
Comment (linjiX) [inline graviti/operation/frame.py:23]: lower case `do` is ok, this is a function name, not a constant member

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/openapi/sheet.py [hunk 12]
```diff
@@ -68,14 +71,13 @@ def create_sheet(
...
         ...     name = "val",
-        ...     schema = "{\"imports\": [{\"repo\":
-        ...         \"https://github.com/Project-OpenBytes/standard@main\", \"types\":
-        ...         [{\"name\": \"label.Category\"}, {\"name\": \"tensor.Image\"}]}],
-        ...         \"type\": \"record\", \"fields\": [{\"name\": \"image\", \"type\":
-        ...         \"tensor.Image\", \"dtype\": \"uint8\", \"shape\": [28, 28]}, {\"name\":
-        ...         \"category\", \"type\": \"label.Category\", \"categories\": [0, 1, 2, 3,
-        ...         4, 5, 6, 7, 8, 9]}]}",
-        ...     _arrow_schema = ""
+        ...     schema = '{"imports": [{"repo": "https://github.com/Project-OpenBytes/standard@\
```
Comment (linjiX) [inline graviti/openapi/sheet.py:74]: shall we update the example in another PR?

Comment (graczhual) [reply graviti/openapi/sheet.py:74]: ok

File: graviti/dataframe/frame.py [hunk 13]
```diff
@@ -90,6 +91,7 @@ def __init__(
 
         self._columns = {}
         self._column_names = []
+        self.operations: List[Union[DataPatch, DataPost]] = []
```
Comment (linjiX) [inline graviti/dataframe/frame.py:94]: add a base class for Operation

File: graviti/operation/frame.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        data: The data for updating.
+        offset: The starting position of the update data.
+        order_by: The string of column names whose order determines the precedence of the sort.
+            Different column names are seperated by ``|``. Multiple indexes can be expressed using
+            dots. If the value is None or incomplete, the remainder will be sorted by the primary
+            key of the sheet.
+
+    """
+
+    DO = staticmethod(update_data)
```
Comment (linjiX) [inline graviti/operation/frame.py:75]: remove this useless code

File: graviti/operation/sheet.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        create_sheet(
+            access_key,
+            url,
+            owner,
+            dataset,
+            draft_number=draft_number,
+            name=self.sheet,
+            schema="",
+            _avro_schema="",
```
Comment (linjiX) [inline graviti/operation/sheet.py:67]: `self.schema`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/operation/frame.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+        Arguments:
+            access_key: User's access key.
+            url: The URL of the graviti website.
+            owner: The owner of the dataset.
+            dataset: Name of the dataset, unique for a user.
+            draft_number: The draft number.
+            sheet: The sheet name.
+
+        """
```
Comment (linjiX) [inline graviti/operation/frame.py:36]: raise `NotImplementError` here

File: graviti/operation/frame.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+class UpdateData(DataFrameOperation):
+    """This class defines the operation that updates the data of a DataFrame.
+
+    Arguments:
+        data: The data for updating.
+
+    """
+
+    def __init__(self, data: List[Dict[str, Any]]) -> None:
```
Comment (linjiX) [inline graviti/operation/frame.py:95]: the data should be a DataFrame rather than List

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/frame.py [hunk 18]
```diff
@@ -91,6 +92,7 @@ def __init__(
 
         self._columns = {}
         self._column_names = []
+        self.operations: List[DataFrameOperation] = []
```
Comment (linjiX) [inline graviti/dataframe/frame.py:95]: shall we add the `opeartions` in `DataFrame.extend`?

File: graviti/manager/draft.py [hunk 19]
```diff
@@ -68,6 +70,17 @@ def __init__(
         self.created_at = created_at
         self.updated_at = updated_at
         self.description = description
+        self.operations: List[SheetOperation] = []
+
+    def __setitem__(self, key: str, value: DataFrame) -> None:
+        if key in self:
+            self.operations.append(DeleteSheet(key))
+        self.operations.append(CreateSheet(key, value.schema))
```
Comment (linjiX) [inline graviti/manager/draft.py:78]: the schema should be copied

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/manager/draft.py [hunk 20]
```diff
@@ -68,6 +70,20 @@ def __init__(
         self.created_at = created_at
         self.updated_at = updated_at
         self.description = description
+        self.operations: List[SheetOperation] = []
+
+    def __setitem__(self, key: str, value: DataFrame) -> None:
+        whether_replace = False
```
Comment (linjiX) [inline graviti/manager/draft.py:76]: `is_replace` is ok

File: graviti/manager/draft.py [hunk 21]
```diff
@@ -68,6 +70,20 @@ def __init__(
...
+        if key in self:
+            whether_replace = True
+        super().__setitem__(key, value)
```
Comment (linjiX) [inline graviti/manager/draft.py:79]: better to add some blank lines for better reading.

File: graviti/manager/draft.py [hunk 22]
```diff
@@ -163,13 +179,31 @@ def commit(self, title: str, description: Optional[str] = None) -> Commit:
...
+        for sheet_name, dataframe in self.items():
+            for df_operation in dataframe.operations:
+                df_operation.do(
+                    self._dataset.access_key,
+                    self._dataset.url,
+                    self._dataset.owner,
+                    self._dataset.name,
+                    draft_number=self.number,
+                    sheet=sheet_name,
+                )
```
Comment (linjiX) [inline graviti/manager/draft.py:206]: shall we clear the data after uploading?

File: graviti/operation/sheet.py [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+        schema: The schema of the DataFrame.
+
+    """
+
+    def __init__(self, sheet: str, schema: PortexType) -> None:
+        super().__init__(sheet)
+        self.schema = schema.to_yaml()
+        pyarrow_schema = pa.shcema(schema.fields.to_pyarrow())  # type: ignore[attr-defined]
+        self._arrow_schema = base64.encodebytes(pyarrow_schema.serialize().to_pybytes())
+        self._avro_schema = json.dumps(convert_arrow_schema_to_avro(pyarrow_schema))
```
Comment (linjiX) [inline graviti/operation/sheet.py:68]: `pyarrow_schema` is `pa.StructType`
`convert_arrow_schema_to_avro` accepts `pa.Schema` will it work?

File: graviti/operation/sheet.py [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+        sheet: The sheet name.
+        schema: The schema of the DataFrame.
+
+    """
+
+    def __init__(self, sheet: str, schema: PortexType) -> None:
+        super().__init__(sheet)
+        self.schema = schema.to_yaml()
+        pyarrow_schema = pa.shcema(schema.fields.to_pyarrow())  # type: ignore[attr-defined]
+        self._arrow_schema = base64.encodebytes(pyarrow_schema.serialize().to_pybytes())
```
Comment (linjiX) [inline graviti/operation/sheet.py:67]: the pyarrow and avro schema should be calculated in `__call__` not `__init__`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/operation/sheet.py [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+    Arguments:
+        sheet: The sheet name.
+        schema: The schema of the DataFrame.
+
+    """
+
+    def __init__(self, sheet: str, schema: PortexType) -> None:
+        super().__init__(sheet)
+        self.schema = schema.to_yaml()
+        pyarrow_schema = pa.shcema(schema.fields.to_pyarrow())  # type: ignore[attr-defined]
```
Comment (linjiX) [inline graviti/operation/sheet.py:66]: `pa.schema`?

**Review by linjiX — APPROVED** [review]

File: graviti/operation/frame.py [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+
+        """
+        add_data(
+            access_key,
+            url,
+            owner,
+            dataset,
+            draft_number=draft_number,
+            sheet=sheet,
+            data=self.data.to_pylist(),
```
Comment (linjiX) [inline graviti/operation/frame.py:89]: what if data is too large?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/operation/frame.py [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+            access_key: User's access key.
+            url: The URL of the graviti website.
+            owner: The owner of the dataset.
+            dataset: Name of the dataset, unique for a user.
+            draft_number: The draft number.
+            sheet: The sheet name.
+
+        """
+        data = self.data.to_pylist()
+        for position in range(0, len(data), _MAX_BATCH_SIZE):
```
Comment (linjiX) [inline graviti/operation/frame.py:85]: we have a method named `chunked` in `tensorbay/utility/itertools.py`, can we reuse it?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Review by AChenQ — APPROVED** [review]

**Pull request merged**